coreax/inverses.py

```diff
@@ -0,0 +1,312 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Classes and associated functionality to approximate regularised inverses of matrices.
+
+Given a matrix :math:`A\in\mathbb{R}^{n\times n}` and some regularisation parameter
+:math:`\lambda\in\mathbb{R}_{\ge 0}`, the regularised inverse of :math:`A` is
+:math:`(A + \lambda I_n)^{-1}` where :math:`I_n` is the :math:`n\times n` identity
+matrix.
+
+When a dataset is very large, methods which have to invert regularised kernel matrices
+evaluated on all of the data can become prohibitively expensive. To reduce this
+computational cost, such methods can instead be approximated (providing suitable
+approximation error can be achieved).
+
+The :class:`RegularisedInverseApproximator` in this module provides the functionality
+required to approximate the inverse of regularised matrices. Furthermore, this class
+allows for "block-inversion" where given an invertible matrix :math:`B`, the block
+array
+
+.. math::
+    A = \begin{bmatrix}
+        B & 0 & \dots & 0 \\
+        0 & 0 & \dots & 0 \\
+        \vdots & \ddots & \dots & \vdots \\
+        0 & 0 & \dots & 0
+    \end{bmatrix}
+
+where only the top-left block contains non-zero elements can be "inverted" to give
+
+.. math::
+    A^{-1} := \begin{bmatrix}
+        B^{-1} & 0 & \dots & 0 \\
+        0 & 0 & \dots & 0 \\
+        \vdots & \ddots & \dots & \vdots \\
+        0 & 0 & \dots & 0
+    \end{bmatrix}.
+
+This functionality allows iterative coreset algorithms which require inverting growing
+arrays to have fully static array shapes and thus be JIT-compilable in JAX.
+
+The most efficient way to compute these "inverses" in JAX requires the `identity` array
+passed in :meth:`~coreax.inverses.RegularisedInverseApproximator.approximate to be a
+matrix of zeros except for ones on the diagonal up to the dimension of the non-zero
+block.
+"""
+
+from abc import abstractmethod
+from typing import Optional
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import jax.random as jr
+from jax import Array
+from typing_extensions import override
+
+from coreax.util import KeyArrayLike
+
+
+class RegularisedInverseApproximator(eqx.Module):
+    """
+    Base class for approximation methods to invert regularised arrays.
+
+    When a dataset is very large, computing the regularised inverse of an array
+     can be very time-consuming. Instead, this property can be approximated by
+    various methods. :class:`RegularisedInverseApproximator` is the base class
+    for implementing these approximation methods.
+
+    :param random_key: Key for random number generation
+    """
+
+    random_key: KeyArrayLike
+
+    @abstractmethod
+    def approximate(
+        self,
+        array: Array,
+        regularisation_parameter: float,
+        identity: Array,
+    ) -> Array:
+        r"""
+        Approximate inverse of regularised array.
```

Comment:
On the regularisation point, what we are doing here is solving (or approximating) a regularised least squares problem (https://en.wikipedia.org/wiki/Regularized_least_squares), it is just that in this case the solution is a approximately the inverse of our target array. So I could rewrite the docstring to reflect that fact more clearly.

On the `approximate` front, the code returns an array that exactly solves the regularised least-square problem above, in the sense that it has the lowest L2 error like you say, but this doesn't mean that you get an exact inverse out at the end. So we'd have to pin down what exactly we mean by approximate, I meant it in the sense that the final inverse is an approximation. Because our target arrays are going to be kernel matrices here which are generally ill-conditioned you can obviously get very "approximate" solutions, if you are not careful.

Reply:
Thanks for looking into this more thoroughly, I will try incorporate the gist of what you have found into the docstrings and argue the case for why we need an approximation.